Feat(serverless-contracts): add SQS contract

CorentinDoue opened the pull request:
# SQS Contract

SQS is not easy to use. Messages are serialized and batched. This contract aims to simplify the usage of SQS. It abstracts serialization and deserialization of the messages while providing type safety at runtime and in the IDE. It also abstracts technical batching process to write or read list of messages.

## Features

- `getSendMessage`: returns a function that **validates**, **serializes** and **send** a single message through SQS
- `getSendMessages`: returns a function that **validates**, **serializes**, **batches** and **send** a list of messages through SQS. Handle **retry** if the SQS is throttled because there are too many messages
- `getHandler`: returns a type safe lambda handler that **deserializes**, **validates**, **handle batched messages one by one** and **properly inform SQS of failed messages**. The callback (internal handler) receives one message and will be called in parallel for each message of the batch. If it throws, the lambda will not throw and only its message will be retuned to the SQS.
- `getLambdaHandler` (naming is bad, but is the same as AGW contract): returns a type safe lambda handler that **deserializes** and **validates** messages. The callback (internal handler) receives the whole batch of messages and is free to process them as it wants.

## TODO before merge

- [x] Publish alpha: https://www.npmjs.com/package/@swarmion/serverless-contracts/v/0.32.0-alpha.0
- [ ] e2e test the alpha on some projects
- [x] Add documentation: https://docs-preview-996--swarmion.netlify.app/docs/how-to-guides/use-serverless-contracts/sqs/

CorentinDoue commented:
> ## [![Quality Gate Failed](https://camo.githubusercontent.com/5333b8e2ff49bd5efdc6d2efd30ec08d7ac8d67c5e80190d03b6c2cb29a5de8b/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f5175616c6974794761746542616467652f71672d6661696c65642d323070782e706e67)](https://sonarcloud.io/dashboard?id=swarmion&pullRequest=996) **Quality Gate failed**
> Failed conditions ![](https://camo.githubusercontent.com/ee78c91cde137cf4668f72b8e1247dadaaeb117c99c74bed0f21b1681b517491/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f6661696c65642d313670782e706e67) [77.1% Coverage on New Code](https://sonarcloud.io/component_measures?id=swarmion&pullRequest=996&metric=new_coverage&view=list) (required ≥ 80%)
> 
> [See analysis details on SonarCloud](https://sonarcloud.io/dashboard?id=swarmion&pullRequest=996)

![image](https://github.com/swarmion/swarmion/assets/31917261/172d0499-130b-4a21-800f-dd4ff908ed78)
![image](https://github.com/swarmion/swarmion/assets/31917261/9e7a7370-669b-4d66-96a7-9eea566b7548)

The coverage is destroyed by the types' folder/files that contains only types. Do you want / Can you exclude them from the sonar coverage?

CorentinDoue commented:
@swarmion/core ⚠️ I think your dependabot is becoming crazy. I think all test (linter, types, ...) hits nx remote cache so all dependabots PR are green and it merges all of them.

For example, when I really execute the test-type command I can see that all vitest.config.ts are broken. Probably since https://github.com/swarmion/swarmion/commit/d71fd6174999ee702eb02988882488cc66396f37

![image](https://github.com/swarmion/swarmion/assets/31917261/70de2e94-f11b-4d70-a9c9-b9ff1843da2a)

fargito commented:
@CorentinDoue my first guess would be that the lockfile isn't properly taken into account by Nx to compute its cache key. I'll try and look into this

fargito commented:
@CorentinDoue I checked on `main` without any Nx cache and everything works as expected. I also tried to run tests on this branch with no cache and it breaks. Can you rebase your branch on main? I think it would solve the type issue

fargito commented on user-docs/documentation/docs/how-to-guides/1-use-serverless-contracts/3-sqs.md:
```diff
@@ -0,0 +1,309 @@
+export const main = getHandler(mySqsContract, {
+  ajv,
+  handleBatchedRecords: false,
+})(async ({ Records }) => {
+  Records.forEach(message => {
```
Nit: the AWS API convention is a bit weird with its UpperCamelCase object keys, I would prefer to use `records`, WDYT?

CorentinDoue replied:
I agree, I changed it

fargito commented on user-docs/documentation/docs/how-to-guides/1-use-serverless-contracts/3-sqs.md:
```diff
@@ -0,0 +1,309 @@
+    // write your business logic for one message
+  });
+  // Handle message failure as you want.
+  // Be aware that unhandled failure will make the whole batch of messages aivalable after the visbility timeout.
```
```suggestion
  // Be aware that unhandled failure will make the whole batch of messages available after the visibility timeout,
```

fargito commented on user-docs/documentation/docs/how-to-guides/1-use-serverless-contracts/3-sqs.md:
```diff
@@ -0,0 +1,309 @@
+:::
+
+:::caution
+This only works with the AWS sdk v3
```
```suggestion
This only works with the AWS SDK v3
```

fargito commented on swarmion.code-workspace:
I think these formatting changes will cause a prettier fail, can you run `pnpm format-fix` at the repo root?

CorentinDoue replied:
I removed this commit after fixing the pnpm.lock (cf https://github.com/swarmion/swarmion/pull/996#issuecomment-2041456965)

fargito commented on packages/serverless-contracts/vitest.config.ts:
```diff
@@ -5,6 +5,7 @@ import { configDefaults, defineConfig } from 'vitest/config';
 export default defineConfig({
   plugins: [tsconfigPaths(), codspeedPlugin()],
   test: {
+    mockReset: true,
```
Curious: why is that necessary now?

CorentinDoue replied:
It's not necessary, but I prefer to have the mocks reset before each test. It avoids side effects of one test on another. Otherwise, lots of `beforeEach` are needed to explicitly reset the mocks when some side effects are discovered, but some can be forgotten.

It's especially useful when counting mock calls. You want all the calls of the previous test to have been forgotten

fargito commented on packages/serverless-contracts/src/contracts/cloudFormation/__tests__/fullContractSchema.type.test.ts:
```diff
@@ -2,8 +2,9 @@ import { JSONSchema } from 'json-schema-to-ts';
 
-type Check =
-  FullContractSchemaType<'hello'> extends JSONSchema ? 'pass' : 'fail';
+type Check = FullContractSchemaType<'hello'> extends JSONSchema
+  ? 'pass'
+  : 'fail';
```
I'm a little bit confused by this change, it should disappear with a `pnpm format-fix` at the repo root

CorentinDoue replied:
I removed this commit after fixing the pnpm.lock (cf https://github.com/swarmion/swarmion/pull/996#issuecomment-2041456965)

fargito commented on packages/serverless-contracts/src/contracts/eventBridge/__tests__/fullContractSchema.type.test.ts:
```diff
@@ -9,10 +9,11 @@ const eventBridgeContract = new EventBridgeContract({
-  FullContractSchemaType<typeof eventBridgeContract> extends JSONSchema
-    ? 'pass'
-    : 'fail';
+type Check = FullContractSchemaType<
+  typeof eventBridgeContract
+> extends JSONSchema
+  ? 'pass'
+  : 'fail';
```
I'm a little bit confused by this change, it should disappear with a `pnpm format-fix` at the repo root

CorentinDoue replied:
I removed this commit after fixing the pnpm.lock (cf https://github.com/swarmion/swarmion/pull/996#issuecomment-2041456965)

fargito commented on packages/serverless-contracts/src/contracts/SQS/utils/serializeMessageAttributes.ts:
```diff
@@ -0,0 +1,61 @@
+      );
+  }
+};
+export const serializeMessageAttributes = <
+  MessageAttributes extends Record<string, unknown>,
+  Contract extends SQSContract,
+>(
```
So this only supports a single depth of objects? If that is the case, maybe this should be mentioned in the docs

fargito commented on packages/serverless-contracts/src/contracts/SQS/utils/messagesValidation.ts:
```diff
@@ -0,0 +1,74 @@
+  contract: Contract;
+  message: SqsMessage<Contract>;
+  options: SendMessageBuilderOptions<Contract>;
+  index?: number;
```
Could this be required?

CorentinDoue replied:
No, the`validateMessage` is used in both `sendMessage` and `sendMessages`. In `sendMessage` there is only one message, so there is no `index`. 

`index` is used here only to print better error messages

fargito reviewed (changes requested):
This is awesome! I have a couple questions/suggestions, but this is impressive work

CorentinDoue commented:
Ok, 

> @swarmion/core ⚠️ I think your dependabot is becoming crazy. I think all test (linter, types, ...) hits nx remote cache so all dependabots PR are green and it merges all of them.
> 
> For example, when I really execute the test-type command I can see that all vitest.config.ts are broken. Probably since [d71fd61](https://github.com/swarmion/swarmion/commit/d71fd6174999ee702eb02988882488cc66396f37)
> 
> ![image](https://private-user-images.githubusercontent.com/31917261/318125098-70de2e94-f11b-4d70-a9c9-b9ff1843da2a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI0OTAyMzEsIm5iZiI6MTcxMjQ4OTkzMSwicGF0aCI6Ii8zMTkxNzI2MS8zMTgxMjUwOTgtNzBkZTJlOTQtZjExYi00ZDcwLWE5YzktYjlmZjE4NDNkYTJhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA3VDExMzg1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY3MjY3NGQ1ZjNjOGZkM2Y0MzE0NTMwM2FlYjIxNjE3YmMzNzI4NWE1NzVmODU3YjVkNzhlZTNhM2IwNWU5NTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.3KxEkqg6pHKZYmZwPIH5IcZ9slvIlV2QLJYE5wsq9wQ)
> 
> ➡️ it breaks my PR 😢

UPDATE: It was my PR, it was started a long time ago and when I rebased it on main with the new dependencies pnpm handled pretty bad the conflicts and ended downgrading some dependencies to the version they were when I started the PR. 

➡️ I fixed this with a rebase interactive, stopped at each commit that changed the `pnpm.lock`, for each of them, I reverted the broken lock and reinstalled the dependencies I changed in my commit from the package.json

CorentinDoue commented:
@fargito I fixed the PR and acted on your comments

CorentinDoue commented:
[RFC] I want to e2e test the contract. Actually this could be done with integration tests. I need your advice/help on some points

1. I want to create a new CDK service `services/contracts-tests` that creates some lambdas that use each contract. Each lambda has some integration test that ensure the contract works as expected. **WDYT?**
2. I can deploy and run the test on one of my account for the development purpose but it could be cool to deploy this service on the staging env and run the test after the deployment as it was done with the integration tests of the orchestrator service. I noticed the staging deployment is broken for 5 month and the integration tests are skipped. I think I probably shut down the account where they were deployed. What do you to do with this? Do you want to remove the staging or is there a plan to get back a dedicated account and fix the CD?

fargito reviewed (approved):
Great work! Thanks a lot.
As for the integration tests:
- I agree with the new service. Maybe we can simply rename the `orchestrator` service and use its infrastructure, since it used to be the source of integration tests
- I'm not really sure we need two separate accounts for prod and staging. We could simply run the integration tests on production after merging on main

Anyway, let's continue this discussion in #525 as on a broader scale, it could also serve as a demonstration of cloud-native integration testing good practices.

fargito commented:
@allcontributors please add @CorentinDoue for docs

fargito commented:
I need to fix #1143 before releasing the beta for 0.32.0, but I'll try and do it fast
